### PR TITLE
main: refactor main package

### PR DESCRIPTION
**Pull request by roobre**
Split from #160 

- Make argumentList a smarter object
- Wrap most of the main logic in a run function which returns an error
- Make logger global (for main package only)
- Minor improvements to error messages and formatting

### CODE REVIEW

File: src/kubernetes.go [hunk 1]
```diff
@@ -230,72 +89,74 @@ func main() {
...
-				os.Exit(1)
-			}
 			ksmNodeIP = kubeletNodeIP
 		} else {
-			innerKSMDiscoverer, err := getKSMDiscoverer(logger)
+			innerKSMDiscoverer, err := args.ksmDiscoverer()
 			if err != nil {
-				logger.Errorf("Error getting KSM discoverer: %v", err)
-				os.Exit(1)
+				return fmt.Errorf("Error getting KSM discoverer: %v", err)
```
Comment (invidian) [inline src/kubernetes.go:143]:
```suggestion
				return fmt.Errorf("getting KSM discoverer: %v", err)
```

File: src/kubernetes.go [hunk 2]
```diff
@@ -230,72 +89,74 @@ func main() {
...
 			}
-			ksmDiscoverer := clientKsm.NewDiscoveryCacher(innerKSMDiscoverer, cacheStorage, ttl, logger)
+
+			ksmDiscoverer := ksmclient.NewDiscoveryCacher(innerKSMDiscoverer, cacheStorage, ttl, logger)
 			ksmClient, err := ksmDiscoverer.Discover(timeout)
 			if err != nil {
-				logger.Errorf("Error discovering KSM: %v", err)
-				os.Exit(1)
+				return fmt.Errorf("Error discovering KSM: %v", err)
```
Comment (invidian) [inline src/kubernetes.go:149]:
```suggestion
				return fmt.Errorf("discovering KSM: %v", err)
```

File: src/main.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package main
+
+import (
+	"os"
+
+	sdkArgs "github.com/newrelic/infra-integrations-sdk/args"
+	"github.com/newrelic/infra-integrations-sdk/log"
+)
+
+var logger log.Logger
```
Comment (invidian) [inline src/main.go:10]: Why can't we pass logger explicitly to `run()`?

File: src/kubernetes.go [hunk 4]
```diff
@@ -204,23 +67,19 @@ func main() {
...
-		logger.Errorf("cluster_name argument is mandatory")
-		os.Exit(1)
+		return fmt.Errorf("cluster_name argument is mandatory")
 	}
 
 	nodeName := os.Getenv(nodeNameEnvVar)
 	if nodeName == "" {
-		logger.Errorf("%s env var should be provided by Kubernetes and is mandatory", nodeNameEnvVar)
-		os.Exit(1)
+		return fmt.Errorf("coudl not find %q in env, which should have been set by Kubernetes", nodeNameEnvVar)
```
Comment (invidian) [inline src/kubernetes.go:77]:
```suggestion
		return fmt.Errorf("could not find %q in env, which should have been set by Kubernetes", nodeNameEnvVar)
```

File: src/kubernetes.go [hunk 5]
```diff
@@ -230,72 +89,74 @@ func main() {
...
-				logger.Errorf("Error getting multiKSM discoverer: %v", err)
-				os.Exit(1)
+				return fmt.Errorf("getting multiKSM discoverer: %v", err)
 			}
-			ksmDiscoveryCache := clientKsm.NewDistributedDiscoveryCacher(ksmDiscoverer, cacheStorage, ttl, logger)
+
+			ksmDiscoveryCache := ksmclient.NewDistributedDiscoveryCacher(ksmDiscoverer, cacheStorage, ttl, logger)
 			ksmClients, err = ksmDiscoveryCache.Discover(timeout)
+			if err != nil {
+				return fmt.Errorf("discovering KSM: %w", err)
```
Comment (invidian) [inline src/kubernetes.go:132]: Perhaps `return fmt.Errorf("discovering distributed KSM: %w", err)`.

File: src/kubernetes.go [hunk 6]
```diff
@@ -395,51 +258,105 @@ func main() {
...
+		// Components will be skipped if their configuration is not correct.
+		if component.Skip {
+			logger.Debugf("Skipping job creation for component %s: %s", component.Name, component.SkipReason)
+			continue
+		}
+
+		componentDiscoverer := controlplaneclient.NewComponentDiscoverer(component, logger, nodeIP, podsFetcher, k8sClient)
+		componentClient, err := componentDiscoverer.Discover(timeout)
+		if err != nil {
+			logger.Errorf("control plane component %s discovery failed: %v", component.Name, err)
```
Comment (invidian) [inline src/kubernetes.go:335]: Should you `continue` here? `componentClient` will most likely be nil, right?

File: src/kubernetes.go [hunk 7]
```diff
@@ -230,72 +89,74 @@ func main() {
 
 	timeout := time.Millisecond * time.Duration(args.Timeout)
 
-	innerKubeletDiscoverer, err := clientKubelet.NewDiscoverer(nodeName, logger)
+	innerKubeletDiscoverer, err := kubeletclient.NewDiscoverer(nodeName, logger)
 	if err != nil {
 		logger.Errorf("Error during Kubelet auto discovering process: %v", err)
```
Comment (invidian) [inline src/kubernetes.go:94]: Should we return error here?

**Comment by invidian** [comment]
@roobre linter is still failing.

**Review by invidian — CHANGES_REQUESTED** [review]
@roobre I had to unresolve conversations which I marked as resolved when I pushed my code to this PR, which has been reverted. It would be nice to get them addressed.

**Review by invidian — APPROVED** [review]

**Comment by roobre** [comment]
Rebased 😭

**Review by invidian — APPROVED** [review]
Nice! :tada:

**Comment by paologallinaharbur** [comment]
@roobre Also this one is becoming stale

**Comment by roobre** [comment]
Closing as with the new architecture in the roadmap, the `main` package will change drastically anyway.

**Comment by invidian** [comment]
Shall we remove the branch as well?

**Pull request closed without merging**